astropy/coordinates/sky_coordinate.py

```diff
@@ -1850,79 +1850,70 @@ def radial_velocity_correction(
         elif self.obstime is not None and self.frame.data.differentials:
             # we do need space motion after all
             coo_at_rv_obstime = self.apply_space_motion(obstime)
-        elif self.obstime is None:
-            # warn the user if the object has differentials set
-            if "s" in self.data.differentials:
-                warnings.warn(
-                    "SkyCoord has space motion, and therefore the specified "
-                    "position of the SkyCoord may not be the same as "
-                    "the `obstime` for the radial velocity measurement. "
-                    "This may affect the rv correction at the order of km/s"
-                    "for very high proper motions sources. If you wish to "
-                    "apply space motion of the SkyCoord to correct for this"
-                    "the `obstime` attribute of the SkyCoord must be set",
-                    AstropyUserWarning,
-                )
+        elif self.obstime is None and "s" in self.data.differentials:
+            warnings.warn(
+                "SkyCoord has space motion, and therefore the specified "
+                "position of the SkyCoord may not be the same as "
+                "the `obstime` for the radial velocity measurement. "
+                "This may affect the rv correction at the order of km/s"
+                "for very high proper motions sources. If you wish to "
+                "apply space motion of the SkyCoord to correct for this"
+                "the `obstime` attribute of the SkyCoord must be set",
+                AstropyUserWarning,
+            )
 
-        pos_earth, v_earth = get_body_barycentric_posvel("earth", obstime)
-        if kind == "barycentric":
-            v_origin_to_earth = v_earth
-        elif kind == "heliocentric":
-            v_sun = get_body_barycentric_posvel("sun", obstime)[1]
-            v_origin_to_earth = v_earth - v_sun
-        else:
+        if kind not in ("barycentric", "heliocentric"):
```

Comment:
This makes the more likely path slower just for error checking, which seems a bad idea. I would keep the logic more similar to what it was (i.e., `if kind=="heliocentric"... elif kind != "barycentric": raise...`). 

The change to inplace subtraction is OK, although I vaguely recall writing explicit `v_earth`, `v_sun` and `v_origin_to_earth` to make the logic clearer.

Reply:
I'm not sure if checking `kind not in ("barycentric", "heliocentric")` slows this function down noticeably, but your suggestion of raising the error in an `elif` is very neat, so I'll apply it regardless.

My idea is that if there is less code then variables tend to be used closer to where they are defined. In this case if `kind == "heliocentric"` then `v_origin_to_earth` gets modified only two lines after it was defined, so it should be very simple to see what the variable contains and introducing an intermediate variable doesn't really help much.

Reply:
Yes, I agree with the general principle and it works well here.